File: woogieon8on/chap8/boj/14501.py
n = int(input())

schedule = []
for _ in range(n):
    schedule.append(list(map(int, input().split())))

profit = [0 for _ in range(n+1)] # i번째 날짜까지의 최대 이익

for i in range(n):
    # 상담 가능한 모든 날짜:
    # (i일 + i번째 상담 시간)부터 마지막 날까지 
    for j in range(i + schedule[i][0], n + 1): 
        if profit[j] < profit[i] + schedule[i][1]: # i번째 상담을 진행했을 때 이익이 더 크다면 상담 진행   
            profit[j] = profit[i] + schedule[i][1]

print(profit[-1]) # 마지막 날짜

File: chae-dahee/ch08/14501.js
const [n, ...arr] = require("fs")
  .readFileSync("chae-dahee/ch08/14501.txt")
  .toString()
  .trim()
  .split("\n");
const N = Number(n);
const counsel = arr.map((tp) => tp.split(" ").map(Number));

const DP = new Array(N).fill(0);

for (let i = 0; i < n; i++) {
  const [period, profit] = counsel[i];
  if (i + period > N) continue;
  DP[i] = DP[i] + profit;

  for (let j = i + period; j < N; j++) {
    DP[j] = Math.max(DP[j], DP[i]);
  }
}
console.log(Math.max(...DP));

//모든 상황을 고려해서 최대 수익을 찾아야 하기 때문
